pcdsdevices/sequencer.py

```diff
@@ -0,0 +1,169 @@
+        # Create our status object
+        return SubscriptionStatus(self.play_status, done, run=True)
+
+    def complete(self):
+        """
+        Complete the EventSequencer's current sequence
+
+        The result of this method varies on the mode that the EventSequencer is
+        configured. If the EventSequencer is either set to "Run Once" or "Run N
+        Times" this method allows the current sequence to complete and returns
+        a status object that indicates a successful completion. However, this
+        mode of operation does not make sense if the EventSequencer is in
+        "Run Forever" mode. In this case, the EventSequencer is stopped
+        immediately and a completed status object is returned.
+
+        Returns
+        -------
+        status: DeviceStatus or SubscriptionStatus
+            Status indicating completion of sequence
+
+        Note
+        ----
+        If you want to stop the sequence from running regardless of
+        configuration use the :meth:`.stop` command.
+
+        """
+        # Clear subscriptions
+        def clear_subs(**kwargs):
+            self.current_step.unsubscribe_all()
+            self.play_count.unsubscribe_all()
+
+        # If we are running forever we can stop whenever
+        if self.play_mode.get() == 2:
+            self.stop()
+            clear_subs()
+            return DeviceStatus(self, done=True, success=True)
+
+        # Otherwise we should wait for the sequencer to end
+        def done(*args, value=None, old_value=None, **kwargs):
+            return value == 0 and old_value == 2
+
+        # Create a SubscriptionStatus
+        st = SubscriptionStatus(self.play_status, done, run=True)
+        st.add_callback(clear_subs)
+        return st
+
+    def collect(self):
+        """
+        Gather information about the state of the Sequencer
+
+        During :meth:`.kickoff`, a monitor is started on the ``current_step``
+        and ``play_count`` signals. The timestamps and values of these signals
+        are stored internally and can be added back into the data stream
+        asynchronously by calling this method.
+        """
+        # Create partial event document
+        event = {'time': time.time(), 'timestamps': dict(), 'data': dict()}
+        for sig, cache in zip((self.current_step, self.play_count),
+                              (self._steps, self._iterations)):
+            event['timestamps'][sig.name] = [evt[0] for evt in cache]
+            event['data'][sig.name] = [evt[1] for evt in cache]
+        # Clear caches
+        self._steps.clear()
+        self._iterations.clear()
+        yield event
+
+    def describe_collect(self):
+        """Describe the collection information"""
+        dd = dict()
+        dd.update(self.play_count.describe())
+        dd.update(self.current_step.describe())
+        return {'sequence': dd}
+
+    def stop(self):
+        """Stop the EventSequencer"""
+        self.play_control.put(0)
```

Comment:
Do you plan to implement `pause` and `resume`? http://nsls-ii.github.io/bluesky/hardware.html#flyer-interface

Reply:
Hmmmmmm.  I am not sure about this. If we do `start` and `stop` we restart the sequence. This could have unintended consequences. I could see it being useful in certain cases though. Requires thought ...

Reply:
You should at least make sure it does the right thing if we ctrl+c during a plan e.g. does the sequence restart when the RE resumes. I don't recall if it calls `kickoff` again or if it just checks for `resume`.

Reply:
`kickoff` is not called again.  What I am suggesting that the sequence will not stop... Doesn't seem right either.

Reply:
I expect the sequence does stop because you have defined `stop`

Reply:
Disagree. `stop_movable_objects` called on `pause` only stops objects that received a `set` command.

https://github.com/NSLS-II/bluesky/blob/a64a205f1b7f090c99ab74074e913ce444c56841/bluesky/run_engine.py#L985

Reply:
I stand corrected. I still think it's weird to keep the sequence running when the scan is paused.